### PR TITLE
8289542: Update JPEG Image Decoding Software to 9e

### PR DESCRIPTION
**Pull request by arapte**
Update libjpeg to current version release [9e](http://www.ijg.org/) of 16-Jan-2022

reference: https://jpegclub.org/reference/reference-sources/

Verified all test run on Windows and MacOS

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8289542](https://bugs.openjdk.org/browse/JDK-8289542): Update JPEG Image Decoding Software to 9e

### Reviewers
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Ajit Ghaisas](https://openjdk.org/census#aghaisas) (@aghaisas - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx pull/874/head:pull/874` \
`$ git checkout pull/874`

Update a local copy of the PR: \
`$ git checkout pull/874` \
`$ git pull https://git.openjdk.org/jfx pull/874/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 874`

View PR using the GUI difftool: \
`$ git pr show -t 874`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/874.diff">https://git.openjdk.org/jfx/pull/874.diff</a>

</details>

### CODE REVIEW

File: modules/javafx.graphics/src/main/native-iio/libjpeg/jdcolor.c [hunk 1]
```diff
@@ -431,22 +431,27 @@ null_convert (j_decompress_ptr cinfo,
...
+  jpeg_component_info *compptr;
 
   while (--num_rows >= 0) {
     /* It seems fastest to make a separate pass for each component. */
-    for (ci = 0; ci < num_comps; ci++) {
+    startptr = *output_buf++;
+    for (ci = 0, compptr = cinfo->comp_info; ci < cinfo->num_components;
+     ci++, compptr++) {
+      if (! compptr->component_needed)
+    continue;        /* skip uninteresting component */
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/native-iio/libjpeg/jdcolor.c:446]: The indentation looks off here.

File: modules/javafx.graphics/src/main/native-iio/libjpeg/jdcolor.c [hunk 2]
```diff
@@ -809,25 +852,35 @@ jinit_color_deconverter (j_decompress_ptr cinfo)
...
-  default:        /* permit null conversion to same output space */
+  default: def_label:    /* permit null conversion to same output space */
     if (cinfo->out_color_space != cinfo->jpeg_color_space)
       /* unsupported non-null conversion */
       ERREXIT(cinfo, JERR_CONVERSION_NOTIMPL);
-    cinfo->out_color_components = cinfo->num_components;
+    i = 0;
+    for (ci = 0; ci < cinfo->num_components; ci++)
+      if (cinfo->comp_info[ci].component_needed)
+    i++;        /* count output color components */
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/native-iio/libjpeg/jdcolor.c:882]: Indentation?

File: modules/javafx.graphics/src/main/native-iio/libjpeg/jdmaster.c [hunk 3]
```diff
@@ -173,20 +171,15 @@ jpeg_calc_output_dimensions (j_decompress_ptr cinfo)
...
-  case JCS_YCCK:
-    cinfo->out_color_components = 4;
-    break;
   default:            /* else must be same colorspace as in file */
-    cinfo->out_color_components = cinfo->num_components;
+    i = 0;
+    for (ci = 0, compptr = cinfo->comp_info; ci < cinfo->num_components;
+     ci++, compptr++)
+      if (compptr->component_needed)
+    i++;    /* count output color components */
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/native-iio/libjpeg/jdmaster.c:181]: intendation?

**Review by kevinrushforth — APPROVED** [review]
Looks good. I tested on all three platforms. I left a couple comments on questionable-looking indentation.

**Comment by arapte** [comment]
> I left a couple comments on questionable-looking indentation.

Hi Kevin, Looks like this is an existing issue, as we replace tabs with 4 spaces.
We have several such existing instances of incorrect layout.

1. https://github.com/openjdk/jfx/blob/5c47295cf7ff9404b3bddaa2c30772fa2eff461c/modules/javafx.graphics/src/main/native-iio/libjpeg/jdmaster.c#L142
2. https://github.com/openjdk/jfx/blob/5c47295cf7ff9404b3bddaa2c30772fa2eff461c/modules/javafx.graphics/src/main/native-iio/libjpeg/jdmaster.c#L439
3. https://github.com/openjdk/jfx/blob/5c47295cf7ff9404b3bddaa2c30772fa2eff461c/modules/javafx.graphics/src/main/native-iio/libjpeg/jdhuff.c#L438

**Comment by kevinrushforth** [comment]
> Looks like this is an existing issue, as we replace tabs with 4 spaces.
> We have several such existing instances of incorrect layout.

Perhaps it would be better to replace TABS with 8 spaces for libjpeg since it looks like that is what their source base assumes? If we were to do that, it would need to be done separately and not part of this update, since we would want to do it consistently for all TABS and not just the ones on lines modified by this update. It may or may not be worth a follow-up cleanup issue.

**Review by aghaisas — APPROVED** [review]

**Pull request closed without merging**